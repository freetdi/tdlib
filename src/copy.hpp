//TODO: header

#ifndef TREEDEC_COPY_HPP
#define TREEDEC_COPY_HPP

#include <boost/graph/copy.hpp>

#ifdef HAVE_GALA_GRAPH_H
#include <gala/boost_copy.h>
#endif

namespace treedec{

using boost::num_vertices;

// copy s to t, but weed out multiedges
// rationale: when boost::copying a directed graph (0,1),(1,0) to an empty
// undirected multigraph, then we get two edges, while one might be more useful
// and more efficient sometimes.
template<class S, class T /* class M */>
void copy_trace(const S& s, T& t /*, M map=identity */)
{
    // TODO template metamagic...
    if(boost::num_vertices(t)){ untested();
        assert(false);
    }else if(!boost::is_multigraph<T>()){
        boost::copy_graph(s, t);
    }else if(boost::is_directed(t) && !boost::is_bidirectional_graph<T>::value) {
        boost::copy_graph(s, t);
    }else{
        t = MOVE(T(boost::num_vertices(s)));
        auto b=boost::edges(s);
        for(; b.first!=b.second; ++b.first){
            typename boost::graph_traits<S>::edge_descriptor f = *b.first;
            auto u=boost::source(f, s);
            auto v=boost::target(*b.first, s);
            if(u<v){
                boost::add_edge(u, v, t);
            }else{
            }
        }
    }
}

} //namespace treedec

#endif

// vim:ts=8:sw=4:et:
