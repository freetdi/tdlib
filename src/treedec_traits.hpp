// Felix Salfelder, 2016
//
// (c) 2016 Goethe-Universit√§t Frankfurt
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//
//   traits for tdlib treedecs.
//
//

#ifndef TD_TREEDEC_TRAITS_HPP
#define TD_TREEDEC_TRAITS_HPP

#include "graph_traits.hpp"

namespace treedec{ //

namespace detail{ //

template<class B, class T, class V>
struct tmpbaghack{ //
    static typename treedec_traits<T>::bag_type& get_bag(T& t, V& v)
    {
        return t[v];
    }
    static typename treedec_traits<T>::bag_type const& get_bag(T const& t, V const& v)
    {
        return t[v];
    }
};

template<class T_t, class V>
struct tmpbaghack<bag_t, T_t, V>{ //
    static typename treedec_traits<T_t>::bag_type& get_bag(T_t& t, V& v)
    {
        return t[v].bag;
    }
    static typename treedec_traits<T_t>::bag_type const& get_bag(T_t const& t, V const& v)
    {
        return t[v].bag;
    }
};
} //detail

template<typename T_t>
inline typename treedec_traits<T_t>::bag_type& bag(
	const typename boost::graph_traits<T_t>::vertex_descriptor& v,
        T_t& T)
{
    typedef typename T_t::vertex_property_type b; //>::bag_type b;
    return detail::tmpbaghack<b,T_t,const typename boost::graph_traits<T_t>::vertex_descriptor&>::get_bag(T, v);
}

template<typename T_t>
inline typename treedec_traits<T_t>::bag_type const& bag(
        const typename boost::graph_traits<T_t>::vertex_descriptor& v,
        T_t const& T)
{
    typedef typename T_t::vertex_property_type b; //>::bag_type b;
    return detail::tmpbaghack<b,T_t,const typename boost::graph_traits<T_t>::vertex_descriptor&>::get_bag(T, v);
}

template<class V, class G>
size_t bag_size(V const & v, G const& g)
{
    return bag(g, v).size();
}

} //treedec

#endif
// vim:ts=8:sw=4:et
