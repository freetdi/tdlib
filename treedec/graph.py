# (c) 2021 Felix Salfelder
# GPLv3+

from . import _graph
from ._graph import _balvvd
