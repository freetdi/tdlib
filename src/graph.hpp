// Felix Salfelder, 2015 - 2016
//
// (c) 2016 Goethe-Universit√§t Frankfurt
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//

/*
 * Graph routines related to tree decompositions.
 *
 * Implemented for generic BGL graphs. Possibly (much) faster for particular
 * graphs.
 *
 * template<typename nIter_t, typename G_t>
 *    void make_clique(nIter_t nIter, G_t &G)
 *
 * template<typename B_t, typename nIter_t, typename G_t>
 *    void copy_neighbourhood(B_t &B, nIter_t nIter, G_t &G)
 *
 * template<class G>
 * void detach_neighborhood(typename boost::graph_traits<G>::vertex_descriptor& c,
 *    G& g, typename outedge_set<G>::type& bag)
 */

#ifndef TD_GRAPH_H
#define TD_GRAPH_H

#include <boost/graph/adjacency_list.hpp>

#include "degree.hpp"
#include "error.hpp"
#include "graph_traits.hpp"
#include "treedec_traits.hpp"
#include "platform.hpp"
#include "random_generators.hpp"
#include "trace.hpp"
#include "error.hpp"

#ifndef HAVE_MYBOOL
#define HAVE_MYBOOL
class mybool{ //
public:
	mybool() : value_(bool())
	{
	}
	/* explicit */ mybool(bool const& t): value_(t) {}
	// /* explicit */ operator bool&() { return value_; }
	/* explicit */ operator bool() const { return value_; }
private:
	char value_;
};
#endif

namespace treedec{ //

#define vertex_iterator_G typename boost::graph_traits<G>::vertex_iterator
#define vertex_descriptor_G typename boost::graph_traits<G>::vertex_descriptor
#define adjacency_iterator_G typename boost::graph_traits<G>::adjacency_iterator

template<typename G>
void remove_vertex(vertex_iterator_G u, G &g)
{
    remove_vertex(*u, g);
}

//Vertex v will remain as isolated node.
//Calls cb on neighbors if degree drops by one,
//before dg will drop.
// TODO: pass edge or edge iterator.
template<typename G>
void contract_edge(vertex_descriptor_G v,
                   vertex_descriptor_G target,
                   G &g,
                   bool /*erase*/=false,
                   vertex_callback<vertex_descriptor_G>* cb=NULL)
{
    adjacency_iterator_G I, E;
    for(boost::tie(I, E)=boost::adjacent_vertices(v, g); I!=E; ++I){
        assert(boost::edge(v, *I, g).second);
        if(*I != target){
            bool added=boost::add_edge(target, *I, g).second;
            if(added){
                //rebasing edge from I-v to I-target.
            }else if(cb){
                //did not add, degree will drop by one.
                (*cb)(*I);
            }
        }else{
        }
    }

    boost::clear_vertex(v, g);
    assert(!boost::edge(v, target, g).second);
}

//Vertex v will remain as isolated node, unless erase.
//If erase, v will be deleted (not collapsed).
template<typename G>
inline void contract_edge(vertex_iterator_G v,
                   vertex_descriptor_G into,
                   G &g,
                   bool erase=false,
                   vertex_callback<vertex_descriptor_G>* cb=NULL)
{ untested();
    contract_edge(*v, into, g, erase, cb);
    if(erase){untested();
        remove_vertex(v, g);
    }
}

} // noboost

namespace treedec{
// turn vertex range into clique.
// call cb on newly created edges and incident vertices.
// returns the number of newly created edges.
template<typename B, typename E, typename G_t>
size_t // hmm
//typename boost::graph_traits<G_t>::edges_size_type
make_clique(B nIt1, E nEnd, G_t &G, typename treedec::graph_callback<G_t>* cb=NULL)
{
    typedef typename boost::graph_traits<G_t>::edge_descriptor edge_descriptor;
    size_t counter=0;
    B nIt2;
    for(; nIt1!=nEnd; ++nIt1){
#if 1
        if(cb){
            (*cb)(*nIt1); // hmm
        }else{
        }
#endif
        nIt2 = nIt1;
        nIt2++;
        for(; nIt2 != nEnd; nIt2++){
            std::pair<edge_descriptor, bool> ep=boost::add_edge(*nIt1, *nIt2, G);
            if(ep.second){
               // new edge created
               ++counter;

               if(cb){
#if 0 // doesn't work. removing center...?
            (*cb)(*nIt1); // hmm
            (*cb)(*nIt2); // hmm
#endif
                   (*cb)(*nIt1, *nIt2);
               }else{
               }
            }else{
            }
            if(!boost::is_directed(G)){
            }else if( boost::edge(*nIt2, *nIt1, G).second ){
                // change later.
            }else{
                boost::add_edge(*nIt2, *nIt1, G);
            }
        }
    }
    return counter;
}

template<typename nIter_t, typename G_t>
void make_clique(nIter_t nIter, G_t &G, treedec::graph_callback<G_t>* cb=NULL)
{
    typename boost::graph_traits<G_t>::adjacency_iterator nIt1, nEnd;
    boost::tie(nIt1, nEnd) = nIter;
    make_clique(nIt1, nEnd, G, cb);
}

// insert the neighbors of v in G into B
template<typename B_t, typename V_t, typename G_t>
void insert_neighbours(B_t &B, V_t v, G_t const &G)
{
    typename boost::graph_traits<G_t>::adjacency_iterator nIt, nEnd;
    boost::tie(nIt, nEnd) = boost::adjacent_vertices(v, G);
    for(; nIt!=nEnd; ++nIt){
        B.insert(*nIt);
    }
}

template<typename B_t, typename V_t, typename G_t>
void insert_neighbours(B_t &B, V_t v, V_t w, G_t const &G)
{
    insert_neighbours(B, v, G);
    insert_neighbours(B, w, G);
}

template<typename B_t, typename V_t, typename G_t>
void insert_neighbours(B_t &B, V_t v, V_t w, V_t x, G_t const &G)
{ untested();
    insert_neighbours(B, v, w, G);
    insert_neighbours(B, x, G);
}

// insert the neighbors of v in G into B
// B starts empty.
// equivalent to B = outedge_set(v, G) where applicable
template<typename B_t, typename V_t, typename G_t>
void assign_neighbours(B_t &B, V_t v, G_t const &G)
{
    typename boost::graph_traits<G_t>::adjacency_iterator nIt, nEnd;
    boost::tie(nIt, nEnd) = boost::adjacent_vertices(v, G);
    for(; nIt!=nEnd; ++nIt){
        B.insert(B.end(), *nIt);
    }
}

// insert neighbors of both vertices, after clearing B
template<typename B_t, typename V_t, typename G_t>
void assign_neighbours(B_t &B, V_t v, V_t w, V_t x, G_t const &G)
{
    B.clear();
    insert_neighbours(B, v, w, G);
    insert_neighbours(B, x, G);
}

// find a minimum degree vertex using linear search.
template <typename G_t>
inline typename boost::graph_traits<G_t>::vertex_descriptor
   get_min_degree_vertex(const G_t &G, bool ignore_isolated_vertices=false)
{
    unsigned int min_degree = UINT_MAX;

    typename boost::graph_traits<G_t>::vertex_iterator vIt, vEnd;
    boost::tie(vIt, vEnd) = boost::vertices(G);
    typename boost::graph_traits<G_t>::vertex_descriptor min_vertex = *vIt++;
    for(; vIt != vEnd; vIt++){
        unsigned int degree = boost::degree(*vIt, G);
        if(degree <= min_degree){
            if(ignore_isolated_vertices && degree == 0){ continue; }
            min_degree = degree;
            min_vertex = *vIt;
        }
    }

    return min_vertex;
}

template <typename G_t>
inline typename boost::graph_traits<G_t>::vertex_descriptor
   get_least_common_vertex(const typename boost::graph_traits<G_t>::vertex_descriptor &min_vertex,
           const G_t &G);

//Copy vertices of G into degree_sequence, ordered by degree, starting with lowest.
template <typename G_t>
inline void make_degree_sequence(const G_t &G,
          std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> &degree_sequence)
{
    unsigned int max_degree = 0;
    typename boost::graph_traits<G_t>::vertex_iterator vIt, vEnd;
    for(boost::tie(vIt, vEnd) = boost::vertices(G); vIt != vEnd; vIt++){
        unsigned int degree = boost::degree(*vIt, G);
        max_degree = (degree>max_degree)? degree : max_degree;
    }

    std::vector<std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> > buckets(max_degree+1);
    for(boost::tie(vIt, vEnd) = boost::vertices(G); vIt != vEnd; vIt++){
        unsigned int degree = boost::degree(*vIt, G);
        if(degree > 0){
            buckets[degree].push_back(*vIt);
        }
    }
    for(unsigned int i = 1; i <= max_degree; i++){
        for(unsigned int j = 0; j < buckets[i].size(); j++){
            degree_sequence.push_back(buckets[i][j]);
        }
    }
}

// obsolete?
template<class G>
struct outedge_set;

template <typename G_t>
std::pair<typename boost::graph_traits<typename graph_traits<G_t>::directed_overlay>::vertex_descriptor,
          typename boost::graph_traits<typename graph_traits<G_t>::directed_overlay>::vertex_descriptor>
    make_digraph_with_source_and_sink(G_t const &G, std::vector<bool> const &disabled,
                 unsigned num_dis,
                 typename graph_traits<G_t>::directed_overlay& dg,
                 std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> &idxMap,
                 typename std::set<typename boost::graph_traits<G_t>::vertex_descriptor> const &X,
                 typename std::set<typename boost::graph_traits<G_t>::vertex_descriptor> const &Y);


#if 0 // not in development?
namespace detail{ //
template<class B, class T, class V>
struct tmpbaghack{ //
    static typename treedec_traits<T>::bag_type& get_bag(T& t, V& v)
    {
        return t[v];
    }
    static typename treedec_traits<T>::bag_type const& get_bag(T const& t, V const& v)
    {
        return t[v];
    }
};

template<class T_t, class V>
struct tmpbaghack<bag_t, T_t, V>{ //
    static typename treedec_traits<T_t>::bag_type& get_bag(T_t& t, V& v)
    {
        return t[v].bag;
    }
    static typename treedec_traits<T_t>::bag_type const& get_bag(T_t const& t, V const& v)
    {
        return t[v].bag;
    }
};
} //detail

template<typename T_t>
inline typename treedec_traits<T_t>::bag_type& bag(
	const typename boost::graph_traits<T_t>::vertex_descriptor& v,
        T_t& T)
{
    typedef typename T_t::vertex_property_type b; //>::bag_type b;
    return detail::tmpbaghack<b,T_t,const typename boost::graph_traits<T_t>::vertex_descriptor&>::get_bag(T, v);
}

template<typename T_t>
inline typename treedec_traits<T_t>::bag_type const& bag(
        const typename boost::graph_traits<T_t>::vertex_descriptor& v,
        T_t const& T)
{
    typedef typename T_t::vertex_property_type b; //>::bag_type b;
    return detail::tmpbaghack<b,T_t,const typename boost::graph_traits<T_t>::vertex_descriptor&>::get_bag(T, v);
}

template<class V, class G>
size_t bag_size(V const & v, G const& g)
{
    return bag(v, g).size();
}
#endif

template<class G_t>
struct deg_chooser { //
    typedef typename misc::DEGS<G_t> type;
    typedef type degs_type; // transition? don't use.
};

// put neighbors of c into bag
// isolate c in g
// return bag
// optionally: pass pointer to bag for storage.
//TODO: not here.
template<class G, class B>
inline void detach_neighborhood(
        typename boost::graph_traits<G>::vertex_descriptor& c,
        G& g, B& bag);

// count number of edges missing in 1-neighborhood of v
template <typename G_t>
inline size_t count_missing_edges(
        const typename boost::graph_traits<G_t>::vertex_descriptor v, G_t const &G)
{
    size_t missing_edges = 0;

    typename boost::graph_traits<G_t>::adjacency_iterator nIt1, nIt2, nEnd;
    for(boost::tie(nIt1, nEnd) = boost::adjacent_vertices(v, G); nIt1 != nEnd; nIt1++){
        nIt2 = nIt1;
        nIt2++;
        for(; nIt2 != nEnd; nIt2++){
            if(!boost::edge(*nIt1, *nIt2, G).second){
                ++missing_edges;
            }
        }
    }
    return missing_edges;
}

// collect neighbors of c into bag
// remove c from graph (i.e. isolate).
// turn subgraph induced by bag into clique.
//
// optionally:
// - call back on each newly created edge, and on each vertex incident to such
//   edge.
// - provide memory through bag pointer.
//
template<class G, class B>
size_t /*G::edge_index_type?*/ make_clique_and_detach(
        typename boost::graph_traits<G>::vertex_descriptor c,
        G& g, B& bag,
        treedec::graph_callback<G>* cb=NULL)
{
    if(boost::is_undirected(g)){
        detach_neighborhood(c, g, bag);
        return make_clique(bag.begin(), bag.end(), g, cb);
    }else{
        /// directed. leave center node untouched.
        // this is a vile hack
        // (and slow for boost::adj_list)
        auto b=boost::adjacent_vertices(c, g);
        return make_clique(b.first, b.second, g, cb);

        auto p=boost::adjacent_vertices(c, g);
        for(; p.first!=p.second; ++p.first)
        {
            // inefficient...
            boost::remove_edge(*p.first, c, g);
        }
    }
}

namespace detail{ //

    // iterate over edges adjacent to both v and s
    // implementation: iterate outedges(v), skip non-outedges(s).
    template<class G>
    class shared_adj_iter : public boost::graph_traits<G>::adjacency_iterator{ //
    public:
        typedef typename boost::graph_traits<G>::adjacency_iterator adjacency_iterator;
        typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
    public: // construct
        shared_adj_iter(adjacency_iterator v, adjacency_iterator ve,
                        vertex_descriptor s, G const& g)
            : adjacency_iterator(v), _ve(ve),
              _s(s), _g(g)
        {
            skip();
        }
        shared_adj_iter(vertex_descriptor v,
                        vertex_descriptor s, G const& g)
            : adjacency_iterator(boost::adjacent_vertices(v, g).first),
              _ve(boost::adjacent_vertices(v, g).second),
              _s(s), _g(g)
        {untested();
            skip();
        }
        shared_adj_iter(const shared_adj_iter& p)
            : adjacency_iterator(p), _ve(p._ve), _s(p._s), _g(p._g)
        {
        }
    public: //ops
        shared_adj_iter& operator++(){
            assert(_ve!=adjacency_iterator(*this));
            assert(adjacency_iterator(*this)!=_ve);
            adjacency_iterator::operator++();
            skip();

            return *this;
        }
    private:
        void skip()
        {
            while(true){
                if(typename boost::graph_traits<G>::adjacency_iterator(*this)==_ve){
                    return;
                }else if(!boost::edge(**this, _s, _g).second){
                    adjacency_iterator::operator++();
                }else{
                    return;
                }
            }
        }
        adjacency_iterator _ve;
        vertex_descriptor _s;
        G const& _g;
    };

}// detail

template<class G>
std::pair<detail::shared_adj_iter<G>, detail::shared_adj_iter<G> >
    inline common_out_edges(typename boost::graph_traits<G>::vertex_descriptor v,
                     typename boost::graph_traits<G>::vertex_descriptor w,
                     const G& g)
{
    typedef typename detail::shared_adj_iter<G> Iter;
    typedef typename boost::graph_traits<G>::adjacency_iterator adjacency_iterator;

    std::pair<adjacency_iterator, adjacency_iterator> p=boost::adjacent_vertices(v, g);
    return std::make_pair(Iter(p.first, p.second, w, g),
                          Iter(p.second, p.second, w, g));
}

// check if two vertices have the same neighborhood.
// return false if not.
template<class G_t>
bool check_twins(typename boost::graph_traits<G_t>::vertex_descriptor v,
                 typename boost::graph_traits<G_t>::vertex_descriptor w,
                 const G_t& G)
{
    typename graph_traits<G_t>::outedge_set_type N1, N2;
    assign_neighbours(N1, v, G);
    assign_neighbours(N2, w, G);

    return(N1==N2);
}

// create ig.
// edge(v, w, ig) == edge(ordering[v], ordering[w], g)
// TODO: inverse map?
// TODO: partial/induced graph?
template<class G>
void immutable_clone(G const &g, typename graph_traits<G>::immutable_type& ig,
   std::vector<typename boost::graph_traits<G>::vertex_descriptor>* ordering=NULL)
{
    typedef typename graph_traits<G>::immutable_type immutable_type;
//    typedef typename graph_traits<G>::vertex_iterator vertex_iterator;
    typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
    typedef typename boost::graph_traits<G>::edge_iterator edge_iterator;

    if(!ordering){ incomplete();
    }
    assert(ordering->size()==boost::num_vertices(g)); // for now.


    if(boost::num_vertices(g) != boost::num_vertices(ig)){ untested();
        // drop a new one... (for now?)
        // inefficient.
        ig = MOVE(immutable_type(boost::num_vertices(g)));
    }else if(boost::num_edges(ig)){ untested();
        for(unsigned i=0; i<boost::num_vertices(g); ++i){
            boost::clear_vertex(i, ig);
        }
    }else{ untested();
    }

    std::vector<unsigned> inverse_ordering(boost::num_vertices(g));
    for(unsigned o=0; o<boost::num_vertices(g); ++o){
        assert((*ordering)[o] < inverse_ordering.size());
        inverse_ordering[(*ordering)[o]] = o;
    }

    for(unsigned o=0; o<boost::num_vertices(g); ++o){
        assert((*ordering)[inverse_ordering[o]] == o);
    }

    edge_iterator e, eend;
    boost::tie(e,eend) = boost::edges(g);
    for(;e!=eend; ++e){
        vertex_descriptor s = boost::source(*e, g);
        vertex_descriptor t = boost::target(*e, g);
        assert(inverse_ordering[s] < boost::num_vertices(ig));
        assert(inverse_ordering[t] < boost::num_vertices(ig));
        assert((s==t) == (inverse_ordering[s]==inverse_ordering[t]));
        boost::add_edge(inverse_ordering[s], inverse_ordering[t], ig);
    }
}

// check if {*vd1, *vd2} is a subset of a bag adjacent to t.
template<class VD_t, class T_t>
class is_in_neighbour_bd{ //
public:
    is_in_neighbour_bd(T_t const& T,
        typename boost::graph_traits<T_t>::vertex_descriptor t)
       : _T(T), _t(t)
    {
    }
public:
    bool operator() (VD_t vd1, VD_t vd2)
    {
        assert(vd1!=vd2);
        if(vd1<vd2){
        }else{ untested();
        }

        typedef typename boost::graph_traits<T_t>::adjacency_iterator bag_iterator;
        bag_iterator nIt, nEnd;
        boost::tie(nIt, nEnd) = boost::adjacent_vertices(_t, _T);
        for(; nIt != nEnd; nIt++){
            BOOST_AUTO(const& ibag, bag(*nIt, _T));

            // BUG, does not work on vectors.
            if(ibag.find(vd1)==ibag.end()){
            }else if(ibag.find(vd2)==ibag.end()){
            }else{
                return true;
            }
        }
        // a = vd1;
        // b = vd2;
        return false;
    }
private: //data
    T_t const &_T;
    typename boost::graph_traits<T_t>::vertex_descriptor _t;
public: // HACK
    long unsigned a, b;
};

// clone subgraph induced by bag into ig.
// store map V(H) -> X \subset V(G) in vdMap
// add more edges for MSVS, TODO: implement properly (how?)
template<class G_t, class T_t, class IG_t, class M_t>
inline void induced_subgraph_with_extra_edges
(G_t const &G, IG_t& ig, // typename graph_traits<G>::immutable_type& ig,
       T_t const& T, typename boost::graph_traits<T_t>::vertex_descriptor bd,
    //   URGHS. no default types without c++11.
     M_t* vdMap /*=NULL*/);


// insert reverse edges.
// dont know how to express "orientation" yet
template<class G>
void make_symmetric(G&g, bool force_oriented=false)
{ untested();
    // assert(g.is_directed);
    auto EE=boost::edges(g);
    if ( /*g.oriented ||*/ force_oriented) { untested();
        for(;EE.first!=EE.second; ++EE.first){
            auto s=boost::source(*EE.first, g);
            auto t=boost::target(*EE.first, g);
            boost::add_edge(t,s,g);
        }
    }else{ untested();
        for(;EE.first!=EE.second; ++EE.first){
            auto s=boost::source(*EE.first, g);
            auto t=boost::target(*EE.first, g);
            if(!boost::edge(s,t,g).second){ untested();
                unreachable(); //?
                boost::add_edge(s,t,g);
            }else if(!boost::edge(t,s,g).second){ untested();
                boost::add_edge(t,s,g);
            }else{ untested();
            }
        }
    }
}
} // treedec


#endif // guard
// vim:ts=8:sw=4:et
