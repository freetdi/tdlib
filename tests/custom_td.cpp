#include <set>
#include <vector>

typedef int var_t;
typedef std::set<unsigned int> sdcc_bagtype;
struct tree_dec_node {
//  typedef boost::vertex_property_tag kind; // do we need it?
	sdcc_bagtype bag;
  std::set<var_t> alive;
  unsigned weight;
};
struct tree_dec_node2 {
//  typedef boost::vertex_property_tag kind; // do we need it?
	std::vector<unsigned> bag;
  std::set<var_t> alive;
  unsigned weight;
};

#include <tdlib/treedec_traits.hpp>

typedef boost::adjacency_list<boost::vecS, boost::vecS,
                              boost::bidirectionalS, tree_dec_node> sbib_tdt;
typedef boost::adjacency_list<boost::vecS, boost::vecS,
                              boost::bidirectionalS, tree_dec_node2> vbib_tdt;

TREEDEC_TREEDEC_BAG_TRAITS(sbib_tdt, bag);
TREEDEC_TREEDEC_BAG_TRAITS(vbib_tdt, bag);

#include <boost/graph/properties.hpp>
#include <boost/graph/graph_utility.hpp>

#include <iostream>
#include <vector>
#include <tuple>
#include <boost/graph/adjacency_list.hpp>
#include <boost/tuple/tuple.hpp>
#include <tdlib/preprocessing.hpp>
#include <tdlib/graph.hpp>
#include <tdlib/thorup.hpp>
#ifdef USE_GALA
#include <tdlib/exact_ta.hpp>
#endif


typedef boost::adjacency_list<boost::vecS, boost::vecS,
		  boost::undirectedS, treedec::bag_t> bag_t_tdt;
typedef boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS,
		  boost::property<treedec::bag_t, std::vector<unsigned> > > prop_tdt;
// this one is used in thorup for example.
typedef boost::adjacency_list<boost::setS, boost::vecS, boost::undirectedS,
		  boost::property<treedec::bag_t, std::set<unsigned int> > > example_tdt;


BOOST_STATIC_ASSERT(
				std::is_same< typename boost::vertex_bundle_type<sbib_tdt>::type,
								tree_dec_node >::value );

#include <boost/graph/copy.hpp>

#if 0
namespace boost{

  template<class U>
  inline void
  put(const put_get_helper<std::vector<U>,
		 property_map<prop_tdt , vertex_all_t >::type  >& pa, unsigned long k,
		 property<treedec::bag_t, std::vector<unsigned> > const& v)
  { untested();
	  auto& b=static_cast<bagstuff::treebagpmap<sbib_tdt> &>(pa)[k];
	  b.clear();
	  for(auto const& i : v.m_value){ untested();
		  treedec::push(b, i);
	  }
  }
  template<class U>
  inline void
  put(const put_get_helper<std::vector<U>,
		 property_map<prop_tdt , vertex_all_t >::type  >& pa, unsigned long k,
		 std::set<unsigned>const& v)
  { untested();
	  auto& b = pa[k];
	  b.clear();
	  for(auto const& i : v){ untested();
		  b.insert(i);
	  }
  }
  template<class U>
  inline void
  put(const put_get_helper<std::vector<U>,
		 property_map<prop_tdt , vertex_all_t >::type  >& pa, unsigned long k,
		 std::vector<unsigned>const& v)
  { untested();
	  auto& b = pa[k];
	  b.clear();
	  for(auto const& i : v){ untested();
		  b.insert(i);
	  }
  }
} // boost
#endif

// too late?
#include <tdlib/thorup.hpp>

int main(int, char**)
{
	sbib_tdt st; // such as used in sdcc
	vbib_tdt vt;
	bag_t_tdt bt; // bundled property, using bag_t
	prop_tdt pt; // bag_t is a property, could be anything.
	example_tdt et;
	auto sv=boost::add_vertex(st);
	boost::add_vertex(pt);

	auto m=boost::get(treedec::bag_t(), st);
	auto xx=boost::get(m, 0);
	boost::get(treedec::bag_t(), st, unsigned(0));
	

	auto sb=boost::get(&tree_dec_node::bag, st, sv);
	auto sb2=boost::get(treedec::bag_t(), st, sv);

	sb.insert(0);

	boost::copy_graph(st, bt);
	boost::copy_graph(bt, st);
	boost::copy_graph(pt, bt);
	boost::copy_graph(et, bt);
	boost::copy_graph(et, st);
//	boost::copy_graph(bt, pt); not yet.
//	boost::copy_graph(st, pt); not yet.
//
	auto n=*boost::vertices(pt).first;
	auto mp=get(treedec::bag_t(), pt);
	auto& b=get(treedec::bag_t(), pt, n);
	treedec::push(b, 1);
	treedec::push(b, 2);
	treedec::push(b, 3);

	n=boost::add_vertex(pt);
	{
		auto& b=get(treedec::bag_t(), pt, n);
		treedec::push(b, 4);
		treedec::push(b, 5);
		treedec::push(b, 6);
	}

	boost::print_graph(pt);
	while(boost::num_vertices(st)){
		boost::remove_vertex(0, st);
	}
	std::cout<<"copy\n";
	boost::copy_graph(pt, st);
	boost::copy_graph(pt, vt);

	auto N=boost::num_vertices(st);
	assert(N==2);
	auto p=boost::vertices(st);

	for(;p.first!=p.second; ++p.first){
		std::cout << *p.first << ":";
		auto q=boost::get(&tree_dec_node::bag, st, *p.first);
		for(auto i : q){
			std::cout << " " << i;
		}
		std::cout <<"\n";
	}

	{
		typedef boost::adjacency_list<boost::vecS, boost::vecS,
				  boost::bidirectionalS > G;
		typedef treedec::thorup<G> alg_B;
		G h;
		alg_B B(h);
		B.do_it();
		auto X=B.get_tree_decomposition();
		boost::copy_graph(X, st);

		B.get_tree_decomposition(st);
	}
	auto const& cpt(pt);
	boost::copy_graph(cpt, st);
	boost::copy_graph(cpt, vt);
	auto const& cbt(bt);
	boost::copy_graph(cbt, st);
//	boost::copy_graph(cbt, vt); not yet
	boost::copy_graph(cbt, st);

}
