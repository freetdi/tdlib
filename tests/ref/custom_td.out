0 <--> 
1 <--> 
copy
0: 1 2 3
1: 4 5 6
elimord
1
0
